### ISSUE TITLE
The result of faiss.Kmeans seems not as good as sklearn.cluster.Kmeans

### ISSUE THREAD

**Issue by fishiwhj** [facebookresearch/faiss#268]
For short text clustering, I compared two libraries: faiss.Kmeans and sklearn.cluster.KMeans, both of which shared the same input representations, but the cluster index assignment for the latter looks much better than the former. 

I also tried tuning some hyper-parameters (num iters) for faiss.Kmeans but did not work. 

I am confused about this, since they use the same clustering algorithm, but why getting different results?

**Comment by mdouze** [comment]
Hi 
That seems weird indeed. What is the sum of squared distances between the training points and the centroids they are assigned to? Is it different between the two implementations?

**Comment by jegou** [comment]
Could you please give more details about the evaluation:
- number and dimensionality of the input vectors
- number of cluster used
- number of iterations

Note, by default Faiss k-means caps the number of points used in the core iteration, typically to a factor times the number of desired clusters. In most applications this probably does not matter much, but I believe that the corresponding parameter `max_points_per_centroid` should be changed if you have relatively imbalance clusters, possibly like in text.

**Comment by fishiwhj** [comment]
@jegou @mdouze 

- The size of the input vectors are 41 * 5709 (41 input vectors with dimension 5709). 

- The iterations are ranged among 25, 50, 100, 200, 500 . 

- The number of clusters are set to 41 -3 = 38, since the input vectors are designed to test the performance of clustering algorithms, only 3 vectors are isolated points. 

The Faiss kmeans algorithms can only find 2/3 of them, while the sklearn kmeans can capture 3/3. The inputs between the two implementations are the same, while the difference between the hyper-parameters of the two implementations are tiny:

In Faiss k-means, 'niter=300', 'cluster_num=38', while others keep default (spherical=False), 

- kmeans = faiss.Kmeans(d, cluster_num, niter, verbose), the API is used as follows:

- kmeans.train(x) 

- D, I = kmeans.index.search(x, 1)

In sklearn k-means, we use the same hyper-parameters as the Faiss one, except some other settings such as 'tol=1e-4' and algorithm='auto'...

- km = KMeans(cluster_num)

- estimated_groups = km.fit_predict(x)

My main concern is faiss kmeans is fail to cluster one of the three isolated point. This is unacceptable since the two input vectors are very similar and should be clustered together. The sklearn kmeans performs well.

**Comment by jegou** [comment]
@fishiwhj: 

Ok, I see. Your evaluation protocol is a bit a corner-case, and not really representative of what k-means is usually intended to optimize in practice, i.e., to minimize the total square loss. By taking a number of clusters being almost as large as the number of points, you have a very atypical setup that mostly depends on the initialization. I would bet that your iterations have essentially no effect and that you converge after 1/3 iterations. Can you confirm?

Scikitlearn uses k-means++ initialization by default (you can also use random points), which is good in the specific corner-case you consider. It should actually gives you perfect result even without any iteration with high probability, because the kind of evaluation you consider is **exactly** what k-means++ has be designed to better handle. 
We have not implemented it in Faiss, because with our former [Yael library,](http://yael.gforge.inria.fr/) which implements both), we observed that the overhead computational cost was not worth the saving (negligible) in all large-scale settings we have considered.  

And looking at the doc, it looks like scikitlearn uses 10 redo by default. Can you confirm that you have specifically ask this when calling it? 

So, if you use scikitlearn with nredo=1 to be fair in the comparison, and possibly use the random points to validate that it is indeed k-means++ that is helping scikit-learn, you should have the same quality with both functions.  

-- rv

**Comment by fishiwhj** [comment]
@jegou 

- I would bet that your iterations have essentially no effect and that you converge after 1/3 iterations. Can you confirm?

Yes it is true. As you said, "It should actually gives you perfect result even without any iteration with high probability". I also tried sklearn with n_init=1, max_iter=1 and the result is the same, but still shows better than the faiss kmeans.

After doing such comparison, I found that in this trivial cases, the faiss kmeans is not worked. It might be worked in some large-scale settings, but It fails to provide the correct clustering results in some "corner-cases".

**Comment by mdouze** [comment]
There is a whole line of research on variants of kmeans and initializations. My general feeling on these works is that they *do* make a difference at a small scale, so maybe it is worthwhile for your use case to look into this literature.

From the Faiss point of view, we just need a fast and robust kmeans implementation that works on datasets of significant size, so in the foreseable future we will not touch the current implementation of k-means.

**Comment by jegou** [comment]
@fishiwhj : unlike what you suggested ("might not work as expected"), Faiss does exactly what a vanilla k-means implementation is asked to do. It is not adapted to all situations: as @mdouze was mentioning, variants of k-means could be better in corner-cases like yours. 
By the way, have you tried Faiss with 10 redos or more? Just increasing the number of redos may be sufficient in your case.
